### DISCUSSION TITLE
KongClusterPlugin changes appended into all ingresses without annotation implemented

### DISCUSSION THREAD

**Discussion by rachelang09** [Kong/kong#7230] [Help] [answered]
### What was created
```
apiVersion: configuration.konghq.com/v1
kind: KongClusterPlugin
metadata:
  name: global-response-transformer
  annotations:
    kubernetes.io/ingress.class: kong
  labels:
    global: "true"
config: 
  add:
    headers:
    - X-Frame-Options: deny
    - Content-Security-Policy: script-src 'self';
    - Strict-Transport-Security: max-age=16070400; includeSubDomains
    - X-Content-Type-Options: nosniff
    - Set-Cookie: HttpOnly; Secure;
plugin: response-transformer
```
### What happened
After creating this plugin, ingresses that are old or newly created will have the headers of the plugin

```
➜  ~ git:(master) ✗ curl new-ingress.com -vvvvv
*   Trying ip-address
* TCP_NODELAY set
* Connected to new-ingress.com(ip-address) port 80 (#0)
> GET / HTTP/1.1
> Host: new-ingress.com
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Content-Length: 17
< Connection: keep-alive
< Date: Wed, 31 Mar 2021 09:17:26 GMT
< X-Frame-Options:  deny
< Content-Security-Policy:  script-src 'self';
< Strict-Transport-Security:  max-age=16070400; includeSubDomains
< X-Content-Type-Options:  nosniff
< Set-Cookie:  HttpOnly; Secure;
```
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: working-ing
  namespace: rachel-test
  annotations:
    kubernetes.io/ingress.class: kong
spec:
  rules:
  - host: new-ingress.com 
    http:
      paths:
      - path: /
        backend:
          serviceName: ingress-class-svc
          servicePort: 80
```
I omitted the ip address and address for security purposes.

According to [kong documentation](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/guides/using-kongclusterplugin-resource/), we need to have `konghq.com/plugin: global-response-transformer` in the `annotations` for it to work. However, it is not reflected properly on my cluster. Initially I thought it was my workspace caching, I asked another colleague to help me test and it's the same.

**Comment by fffonion** [comment]
Hi @rachelang09 , since you create the CRD with `global: "true"` in labels, it's applied globally to all ingresses.

**Comment by rachelang09** [comment]
Hi @fffonion thanks for your reply! I set it to `global: "true"` but it did not apply to my testing ingress until I added in the annotations. However, after that, the rest of the ingresses old or new are appended with it

**Comment by rachelang09** [comment]
After doing up 4 use cases, I realised that for the ingress to automatically append the plugin, we need to have both `annotation: kubernetes.io/ingress.class: kong` and `labels:global: "true"`. Also noted that `annotation: kubernetes.io/ingress.class: kong` only affects ingresses that uses the same `annotation: kubernetes.io/ingress.class: kong` even if they do not have `annotations: konghq.com/plugins: global-response-transformer`

I tried to remove `annotation: kubernetes.io/ingress.class: kong` only leaving `labels:global: "true"` as I presume that by doing so, it will append to all ingresses but it did not happen. 

May I check, where in the [documentation](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/guides/using-kongclusterplugin-resource/#configuring-plugins-on-ingress-resources) states such behaviour as I am unable to find it.

**Comment by fffonion** [comment, marked as answer]
@rachelang09 your obsercation is correct. `annotation: kubernetes.io/ingress.class: kong` describes who will be responsible for the Ingress, by setting so, the kong ingress controller will be able to find ingresses that matches its own ingress.class parameter (which is `"kong"` by default).
While `labels:global: "true"` is a custom protocol Kong Ingress Controller understands, when seeing this label for a KongClusterPlugin CRD, it creates the plugin as a global plugin in Kong. Otherwise it creates the plugin associated with Route that has `annotations: konghq.com/plugins: <name>`.

**Comment by rachelang09** [comment]
@fffonion thanks for your response! I'm just bumped out that it's not stated in the documentation and I had to do up a few use case to find out the actual implementation of the plugin

**Comment by fffonion** [comment]
@rachelang09 No problem. I will see how it can be improved in docs and update here.
